libminifi/src/utils/BaseHTTPClient.cpp

```diff
@@ -160,4 +160,116 @@ std::string URL::toString() const {
   }
 }
 
+/**
+ * Receive HTTP Response.
+ */
+size_t HTTPRequestResponse::receiveWrite(char *data, size_t size, size_t nmemb, void *p) {
+  try {
+    if (p == nullptr) {
+      return CALLBACK_ABORT;
+    }
+    auto *callback = static_cast<HTTPReadCallback *>(p);
+    if (callback->stop) {
+      return CALLBACK_ABORT;
+    }
+    callback->write(data, (size * nmemb));
+    return (size * nmemb);
+  } catch (...) {
+    return CALLBACK_ABORT;
+  }
+}
+
+/**
+ * Callback for post, put, and patch operations
+ * @param data output buffer to write to
+ * @param size number of elements to write
+ * @param nmemb size of each element to write
+ * @param p input object to read from
+ */
+size_t HTTPRequestResponse::send_write(char *data, size_t size, size_t nmemb, void *p) {
+  try {
+    if (p == nullptr) {
+      return CALLBACK_ABORT;
+    }
+    auto *callback = reinterpret_cast<HTTPUploadCallback *>(p);
+    return callback->getDataChunk(data, size * nmemb);
+  } catch (...) {
+    return CALLBACK_ABORT;
+  }
+}
+
+int HTTPRequestResponse::seek_callback(void *p, int64_t offset, int) {
+  try {
+    if (p == nullptr) {
+      return SEEKFUNC_FAIL;
+    }
+    auto *callback = reinterpret_cast<HTTPUploadCallback *>(p);
+    return callback->setPosition(offset);
+  } catch (...) {
+    return SEEKFUNC_FAIL;
+  }
+}
+
+size_t HTTPUploadByteArrayInputCallback::getDataChunk(char *data, size_t size) {
+  if (stop) {
+    return HTTPRequestResponse::CALLBACK_ABORT;
+  }
+  size_t buffer_size = getBufferSize();
+  if (pos <= buffer_size) {
+    size_t len = buffer_size - pos;
+    if (len <= 0) {
+      return 0;
+    }
+    auto *ptr = getBuffer(pos);
+
+    if (ptr == nullptr) {
+      return 0;
+    }
+    if (len > size)
```

Comment:
I would use `std::min` instead

Reply:
Also added to https://issues.apache.org/jira/browse/MINIFICPP-1971.